src/cli_util.rs

```diff
@@ -2028,6 +2028,35 @@ impl ValueParserFactory for RevisionArg {
     }
 }
 
+fn resolve_default_command(config: &config::Config, app: &Command, string_args: &mut Vec<String>) {
+    const PRIORITY_FLAGS: &[&str] = &["help", "--help", "-h", "--version", "-V"];
+
+    let has_priority_flag = string_args
+        .iter()
+        .any(|arg| PRIORITY_FLAGS.contains(&arg.as_str()));
+    if has_priority_flag {
+        return;
+    }
+
+    let app_clone = app
+        .clone()
+        .allow_external_subcommands(true)
+        .ignore_errors(true);
+    let matches = app_clone.try_get_matches_from(string_args.clone()).ok();
+
+    if let Some(matches) = matches {
+        if matches.subcommand_name().is_none() {
+            let default_command = config.get_string("ui.default-command").unwrap_or_else(|_| {
+                eprintln!("HINT: Use `jj -h` for a list of available commands.");
+                eprintln!("Set the config `ui.default-command = \"log\"` to disable this message.");
```

Comment:
Wait, sorry, these errors should use the `Ui` object instead of printing directly to stderr. Something like this:

```
        writeln!(ui.hint(), "Hint: ....");
```

We seem to capitalize only the "H" in "Hint" elsewhere, so please fix that too.